pandas/tests/reshape/merge/test_merge.py

```diff
@@ -354,8 +354,8 @@ def test_merge_join_key_dtype_cast(self):
         df = merge(df1, df2, how="outer")
 
         # GH13169
-        # this really should be bool
-        assert df["key"].dtype == "object"
+        # GH#40073
+        assert df["key"].dtype == "bool"
```

Comment:
@debnathshoham is this fixing a different bug than the EA case also? Probably worth mentioning in the whatsnew / adding a more explicit test if so

Reply:
actually this is fixing the cases wherever `pandas.Index` casts to object. As per the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.Index.html), `Index` takes/infers numpy dtype, but surprisingly is casting `Index[True]` to object (although ``np.array[True].dtype`` comes bool).

I will add a specific to the `bool` case (not sure if I am missing any other dtype `Index` casts to object.